### PR TITLE
Improved go to type definition on type references of global container-like types using allowlist

### PR DESCRIPTION
**Pull request by Andarist**
Implementation-wise this seems to be pretty complete - unless some further refinements for the desired behavior get requested. It probably could use more tests but let's first settle on the desired behavior. I open this to gather initial feedback :)

At the moment, this just adds new entries to the returned list so VS Code is able to display this. Otherwise, there are no labels nor anything to indicate that those new entries are somehow different from the rest of the list.

Showcased go to type definition improvement in VS Code

closes https://github.com/microsoft/TypeScript/issues/14148

cc @andrewbranch

### CODE REVIEW

File: tests/baselines/reference/goToTypeDefinition_arrayType.baseline.jsonc [hunk 1]
```diff
@@ -0,0 +1,4564 @@
+// === goToType ===
+// === /tests/cases/fourslash/goToTypeDefinition_arrayType.ts ===
+// type User = [|{| defId: 0 |}{ name: string }|];
+// declare const users: User[]
+// /*GOTO TYPE*/users
+
+// === lib.d.ts ===
```
Comment (andrewbranch) [inline tests/baselines/reference/goToTypeDefinition_arrayType.baseline.jsonc:7]: @sheetalkamat this baseline shows an unfortunate edge case. I thought we discussed whether less than a full file could be printed back, but I don’t remember the outcome. Is there anything we can do about this?

**Comment by Andarist** [comment]
@andrewbranch added a screenshot to the PR's description

File: tests/baselines/reference/goToTypeDefinition_arrayType.baseline.jsonc (hunk 1 above)
Comment (sheetalkamat) [reply tests/baselines/reference/goToTypeDefinition_arrayType.baseline.jsonc:7]: Could the test include custom lib.d.ts

Comment (andrewbranch) [reply tests/baselines/reference/goToTypeDefinition_arrayType.baseline.jsonc:7]: Can we not just truncate lines +/- _N_ away from markers?

**Comment by andrewbranch** [comment]
@fatcerberus did you briefly forget that `User[]` is actually `Array<User>`? 😄

**Comment by fatcerberus** [comment]
@andrewbranch Hm, I thought that might be what this was but then that didn’t make sense to me either - for `Array` it does but in the general case for arbitrary `Foo<Bar>`, I’m not sure it usually makes sense to jump right to `Bar`.

**Comment by Andarist** [comment]
>  for Array it does but in the general case for arbitrary Foo<Bar>, I’m not sure it usually makes sense to jump right to Bar.

Maybe. This all is an open question here. Is this a good idea in the first place? Could improved IDE's UI help with this use case anyhow? Should all type references with single type params be expanded? Should we only do this for an allowlist of type references (like `Array`, `Promise`, etc)? What about `Promise<Array<User>>` - should this be "unwrapped" recursively?

**Comment by andrewbranch** [comment]
Maybe a list of blessed collection/wrapper types is better. On the other hand, I can think of a _lot_ of non-collection utility types where this is nice: `Partial`, `Require`, Readonly`, even multi-argument ones like `Omit`.

Probably just doing Promise and the iterables is the best place to start. Discussing with the team today.

**Comment by andrewbranch** [comment]
@typescript-bot pack this

**Comment by andrewbranch** [comment]
@mjbvz is going to look into updating VS Code’s API to let us make titled groups of definitions. In the meantime, let’s undo my suggestion of doing this for all single-type-parameter type aliases and have an allow list. Copilot and I came up with this, additions accepted:

- Array
- ArrayLike
- ReadonlyArray
- Promise
- PromiseLike
- Iterable
- IterableIterator
- AsyncIterable
- Set
- WeakSet
- ReadonlySet
- Map?
- WeakMap?
- ReadonlyMap?
- Partial
- Required
- Readonly
- Pick?
- Omit

**Comment by Andarist** [comment]
Cool, should I do this recursively (like in the `Promise<Array<User>>` case)? Do u have any sketch for the proposed VS Code UI? I wonder if i u have considered a tree-like view that would allow expanding each type argument on its own - this way it could be pretty generic, allowing everyone to grab the info they need at any given moment

**Comment by fatcerberus** [comment]
> Copilot and I came up with this

It’s a good list, though I’m now wondering what exactly you asked the AI to get specifically those types…

**Comment by andrewbranch** [comment]
@fatcerberus I just started writing the list and it continued.

@Andarist I would say no on the recursive unwrapping, and if the type argument to one of the things in my list is another thing in the list, then only show the top level? I think it will make more sense to revisit these more complex scenarios when/if there is an API and UI to speak of. The existing issue is https://github.com/microsoft/vscode/issues/74237.

**Comment by Andarist** [comment]
@andrewbranch pushed out a variant based on the given allowlist. Could you take a look if the implementation looks alright to you? If yes then I could add some more tests to this PR.

**Comment by andrewbranch** [comment]
Looks right to me!

File: tests/baselines/reference/goToTypeDefinition_Pick.baseline.jsonc [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+// }|];
+// 
+// /**
+//  * Construct a type with a set of properties K of type T
+// --- (line: 1598) skipped ---
+
+// === /tests/cases/fourslash/goToTypeDefinition_Pick.ts ===
+// type User = { id: number; name: string; };
+// declare const user: Pick<User, "name">
+// /*GOTO TYPE*/user
```
Comment (sheetalkamat) [inline tests/baselines/reference/goToTypeDefinition_Pick.baseline.jsonc:18]: i owuld have expected it to go to "User" type instead of Pick ?

**Comment by Andarist** [comment]
@andrewbranch i pushed out what I managed to put together for this. `Pick` works, different kinds of aliases work now. It ain't exactly overly pretty... but it seems to be functional. Could you take a look at what I crafted here?

File: src/compiler/sys.ts [hunk 3]
```diff
@@ -1578,7 +1578,8 @@ export let sys: System = (() => {
             disableCPUProfiler,
             cpuProfilingEnabled: () => !!activeSession || contains(process.execArgv, "--cpu-prof") || contains(process.execArgv, "--prof"),
             realpath,
-            debugMode: !!process.env.NODE_INSPECTOR_IPC || !!process.env.VSCODE_INSPECTOR_OPTIONS || some(process.execArgv, arg => /^--(inspect|debug)(-brk)?(=\d+)?$/i.test(arg)),
+            // debugMode: !!process.env.NODE_INSPECTOR_IPC || !!process.env.VSCODE_INSPECTOR_OPTIONS || some(process.execArgv, arg => /^--(inspect|debug)(-brk)?(=\d+)?$/i.test(arg)),
+            debugMode: true,
```
Comment (andrewbranch) [inline src/compiler/sys.ts:1582]: replay.io 🥺

**Review by andrewbranch — APPROVED** [review]
Works for me, minus the debugging artifact

File: src/compiler/sys.ts (hunk 3 above)
Comment (jakebailey) [reply src/compiler/sys.ts:1582]: Would really like to get #49027 if we think it's good FWIW

Comment (Andarist) [reply src/compiler/sys.ts]: Eh, it ought to happen after months of keeping this locally and trying my best to not commit this :P

A woman saying 'okay, then you caught me'

**Comment by Andarist** [comment]
@andrewbranch reverted the accidental `debugMode` change

**Review by sheetalkamat — APPROVED** [review]

**Comment by andrewbranch** [comment]
I think I would lean toward waiting until 5.2 opens to merge this, that way if we get feedback that it’s confusing, we may be able to work on the suggested VS Code protocol/UI changes to complement it.

**Comment by Andarist** [comment]
Fine by me! :) Truth to be told, I kinda expected that 😉

**Pull request merged**